glusterd: enhancement around port mapper

nik-redhat opened the pull request:
Issue:
pmap_registry_new() tests bind() for all ports from base_port
to max_port for pmap registry initialization. The initialized
data will be stale when we want to use the ports, so we avoid
keeping any entry of free ports at the beginning and check the
port if it's free or not only before using it.
    
Fix:
1 .Modified the code for the port mapper completely, that is removed
    unwanted code from the file glusterd_pmap.c as now we are not
    using the array for storing the port number, instead now just randomized
    the port number within the base port and max port in the func
    pmap_port_alloc() and then just returning the port number.

2. Changed the code to use a list and not an array to store the port info 
    such as port number, brick, and xprt.

3. Modified the search, bind, remove, and alloc functions to store the 
    port information in a list. 
    
Fixes: #786
    
Signed-off-by: nik-redhat <nladha@redhat.com>

nik-redhat committed:
glusterd: enhancement around port mapper

pmap_registry_new() tests bind() for all ports from base_port
to max_port for pmap registry initialization. The initialized
data will be stale when we want to use the ports, so we avoid
keeping any entry of free ports at the beginning and check the
port if its free or not only before using it.

pmap_registry_alloc() searches for a free port from the base_port
for every volume. Instead, we can start from last_alloc port and
if we don't find a free port between last_alloc port max_port, we
search from base_port to last_alloc port.

Fixes: #786

Signed-off-by: nik-redhat <nladha@redhat.com>

nik-redhat committed:
glusterd: enhancement around port mapper

Issue:
pmap_registry_new() tests bind() for all ports from base_port
to max_port for pmap registry initialization. The initialized
data will be stale when we want to use the ports, so we avoid
keeping any entry of free ports at the beginning and check the
port if its free or not only before using it.

Fix:
1.Modified the code for the port mapper completely, that is removed
unwanted code from the file glusterd_pmap.c as now we are not
using the array for storing the port number, instead randomizing
the port number within the base port and max port in the func
pmap_port_alloc() and then just treturing the port number.

2.Removed few unwanted operations like search, bind, extend, remove
registry as we are not using the array anymore.

3.Modified few required functions so as to work without storing
the results in the array.

Fixes: #786

Signed-off-by: nik-redhat <nladha@redhat.com>

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

nik-redhat committed:
Removed unwanted struct

Signed-off-by: nik-redhat <nladha@redhat.com>

nik-redhat committed:
Fixed clang format

Signed-off-by: nik-redhat <nladha@redhat.com>

mykaul commented on xlators/mgmt/glusterd/src/glusterd-snapd-svc.c:
```diff
@@ -335,7 +335,7 @@ glusterd_snapdsvc_start(glusterd_svc_t *svc, int flags)
     }
 
-    snapd_port = pmap_assign_port(THIS, volinfo->snapd.port, snapd_id);
+    snapd_port = pmap_port_alloc(THIS);
```
```suggestion
    snapd_port = pmap_port_alloc(this);
```

mykaul commented on xlators/mgmt/glusterd/src/glusterd-statedump.c:
```diff
@@ -216,19 +214,6 @@ glusterd_dump_priv(xlator_t *this)
         /* Dump peer details */
         GLUSTERD_DUMP_PEERS(&priv->peers, uuid_list, _gf_false);
 
-        /* Dump pmap data structure from base port to last alloc */
```
nit: we may want to dump the port, where is it done?

nik-redhat replied:
In that case, we will have to dump the ports using the brickinfo struct, by traversing through the brick_list.

nik-redhat committed:
Fix unused variables

Signed-off-by: nik-redhat <nladha@redhat.com>

mykaul commented on xlators/mgmt/glusterd/src/glusterd-utils.c:
```diff
@@ -2070,7 +2070,7 @@ glusterd_volume_start_glusterfs(glusterd_volinfo_t *volinfo,
     }
 
-    port = pmap_assign_port(THIS, brickinfo->port, brickinfo->path);
+    port = pmap_port_alloc(THIS);
```
```suggestion
    port = pmap_port_alloc(this);
```

mykaul commented on xlators/mgmt/glusterd/src/glusterd-utils.c:
```diff
@@ -2167,8 +2167,7 @@ glusterd_volume_start_glusterfs(glusterd_volinfo_t *volinfo,
         }
-        rdma_port = pmap_assign_port(THIS, brickinfo->rdma_port,
-                                     rdma_brick_path);
+        rdma_port = pmap_port_alloc(THIS);
```
```suggestion
        rdma_port = pmap_port_alloc(this);
```

mykaul commented on xlators/mgmt/glusterd/src/glusterd-volume-ops.c:
```diff
@@ -2851,8 +2851,7 @@ glusterd_clearlocks_get_local_client_ports(glusterd_volinfo_t *volinfo,
 
-        port = pmap_registry_search(THIS, brickname, GF_PMAP_PORT_BRICKSERVER,
-                                    _gf_false);
+        port = brickinfo->port;
```
nit: I don't think you need the assignment at all - you use that info only once.

mykaul commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -47,29 +49,25 @@ pmap_port_isfree(int port)
+/* Create a new pmap registry */
+
 static struct pmap_registry *
 pmap_registry_new(xlator_t *this)
```
nit: pass this->private into this function instead of this.

mykaul commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -47,29 +49,25 @@ pmap_port_isfree(int port)
     struct pmap_registry *pmap = NULL;
-    int i = 0;
 
     pmap = CALLOC(sizeof(*pmap), 1);
```
nit: seems like you can GF_MALLOC instead, and just initialize all members?

mykaul commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -47,29 +49,25 @@ pmap_port_isfree(int port)
+/* Fetch the already created pamp struct or create new one */
+
 struct pmap_registry *
 pmap_registry_get(xlator_t *this)
```
nit: I think here as well, you can pass glusterd_conf_t *priv

mykaul commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -489,10 +186,14 @@ __gluster_pmap_brickbyport(rpcsvc_request_t *req)
-    if (!rsp.brick) {
-        rsp.op_ret = -1;
-        rsp.brick = "";
+    ret = glusterd_get_brickinfo(THIS, NULL, args.port, &brickinfo);
```
```suggestion
    ret = glusterd_get_brickinfo(this, NULL, args.port, &brickinfo);
```

mykaul commented on xlators/mgmt/glusterd/src/glusterd-utils.c:
```diff
@@ -7258,7 +7239,7 @@ glusterd_get_brickinfo(xlator_t *this, const char *brickname, int port,
             if (gf_uuid_compare(tmpbrkinfo->uuid, MY_UUID))
                 continue;
-            if (!strcmp(tmpbrkinfo->path, brickname) &&
+            if (!strcmp(tmpbrkinfo->path, brickname) ||
```
not sure why is this change done here, but in any case, perhaps worth doing the port comparison before the strcmp()

nik-redhat replied:
To incorporate the functions glusterd_pmap_portbybrick and glusterd_pmap_bruckbyport to work fine.
Will, make the changes to compare port first.

mykaul reviewed (commented):
Small comments. Overall, exactly the cleanup I've expected to see. Need to test with a large number of volumes and bricks!

nik-redhat commented:
> Small comments. Overall, exactly the cleanup I've expected to see. Need to test with a large number of volumes and bricks!

Cool, will be addressing the comments ASAP.

mykaul commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -89,337 +87,32 @@ pmap_registry_get(xlator_t *this)
-        case GF_PMAP_PORT_BRICKSERVER:
+    while (true) {
+        p = (rand() % (pmap->max_port - pmap->base_port + 1)) + pmap->base_port;
+        if (pmap_port_isfree(p)) {
```
So, here we check if it's free by binding, then *closing* that bind. Why? Keep it open, use it!
(however, note that we deal only with IPv4 in pmap_port_isfree() , which may not be a good overall solution for IPv6 only setups).

I'd fix it in a follow-up patch, though.

raghavendra-talur replied:
The reason for closing it here is because glusterd is not the process which actually wants to bind to it. Here, glusterd is doing the probe on behalf of the gluster brick process, selecting the port and then starting the brick process with the port number as an argument.

@nik-redhat If you are interested in further cleanups, this is certainly a good one in the same code area. Introduce a method where glusterd starts the brick process with a magic port number or no number. In that case, it will be up to the brick process to select the port to bind to. You will have to pass the range of valid ports(base port and max port) though. As you have already updated the pmap_registry in this PR, I don't think any change will be required once the port is selected by the brick process and it binds to it.

nik-redhat replied:
So, the magic port here would be random port only, right? It too finds the port within the base_port and max_port.
And, in the changes to something similar is done. Please pardon me if I understood something wrong. How else shall we do so then?

raghavendra-talur replied:
I created an issue #1784 that has details on the other refactor I suggested above. We can discuss that change there. This PR looks good for what it is aiming for.

nik-redhat replied:
Cool 👍🏼

mykaul commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -444,15 +137,15 @@ __gluster_pmap_portbybrick(rpcsvc_request_t *req)
-    brick = args.brick;
+    ret = glusterd_get_brickinfo(THIS, args.brick, 0, &brickinfo);
+    /* get the port number from the brickinfo struct */
+    if (brickinfo) {
```
nit: clang will complain on ignoring ret from the previous line.

mykaul commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -489,10 +185,14 @@ __gluster_pmap_brickbyport(rpcsvc_request_t *req)
-        rsp.brick = "";
+    ret = glusterd_get_brickinfo(THIS, NULL, args.port, &brickinfo);
+    /* get the brickname from the brickinfo struct */
+    if (brickinfo) {
```
nit - clang will complain on ignoring 'ret' from the previous line.

mykaul commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -582,16 +281,14 @@ __gluster_pmap_signout(rpcsvc_request_t *req)
-                                          GF_PMAP_PORT_BRICKSERVER, req->trans,
-                                          _gf_false);
+
+        rsp.op_ret = 0;
```
I don't think we still support RDMA, do we? I think we might be able to remove this part altogether. 
See https://review.gluster.org/#/c/glusterfs/+/24031/ (which we never merged)

nik-redhat committed:
Changed few nits

Signed-off-by: nik-redhat <nladha@redhat.com>

nik-redhat committed:
Fix smoke

Signed-off-by: nik-redhat <nladha@redhat.com>

nik-redhat committed:
Chnaged to GF_MALLOC

Signed-off-by: nik-redhat <nladha@redhat.com>

nik-redhat commented:
> nit: why pass 'this' here and not 'priv' ?

There is some issue with using the struct glusterd_conf_t in the glusterd-pmap.h file. Not, able to figure out why it can't read the structure declartion, so used this.

nik-redhat committed:
Fix for smoke

Signed-off-by: nik-redhat <nladha@redhat.com>

nik-redhat commented:
> Do we need rdma port?

I left it for the time being, as it might affect other areas as well. It would be good to remove it altogether or the maximum of it at once, rather than removing it here once. What do you suggest?

nik-redhat committed:
Fix for smoke fails

Signed-off-by: nik-redhat <nladha@redhat.com>

mohit84 reviewed (commented):
The patch is touching sensitive code path specific to brick_mux so we need regress testing to backport it.At this point i would not recommend to backport it.

-1

nik-redhat commented:
> The patch is touching sensitive code path specific to brick_mux so we need regress testing to backport it.At this point i would not recommend to backport it.
> 
> -1

The code changes do touch code specific to brick_mux, but I believe that the changes won't affect brick_mux working as it just reduces some additional operations when brick_mux is active. But, to be on the safer side a regress testing would be a better approach.
So, is it something you will be doing it @mohit84 or it shall be done by me following some steps for that?

mykaul commented:
I don't think there's any need for a backport anyway, since this is not a bug fix but an enhancement. 
We definitely need to run brick mux regression on this (as well as regular regression*).

* Would have been nice if our regression tests could capture memory usage of our various processes, to catch leaks (or additional memory consumption) - or in this case, reduced memory consumption.

nik-redhat commented:
The regression test and brick-mux regression needs to be triggered by @mohit84

nik-redhat commented:
@mohit84 can you please run the regression test for the PR?

nik-redhat commented:
@gluster/gluster-maintainers can someone please run the regression tests for this PR?

pranithk commented:
/run regression

xhernandez commented on libglusterfs/src/glusterfs/mem-types.h:
```diff
@@ -134,6 +134,7 @@ enum gf_common_mem_types_ {
     gf_common_mt_server_cmdline_t,     /* used only in one location */
     gf_common_mt_latency_t,
-    gf_common_mt_end
+    gf_common_mt_end,
```
This value must be the last one of the enum.

nik-redhat replied:
Okay, will update that.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -47,40 +49,36 @@ pmap_port_isfree(int port)
     return pmap;
 }
 
+/* Fetch the already created pamp struct or create new one */
```
```suggestion
/* Fetch the already created pmap struct or create new one */
```

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -89,337 +87,32 @@ pmap_registry_get(xlator_t *this)
-        case GF_PMAP_PORT_LEASED:
-        case GF_PMAP_PORT_BRICKSERVER:
+    while (true) {
+        p = (rand() % (pmap->max_port - pmap->base_port + 1)) + pmap->base_port;
```
I agree with this change, but we'll need to document it properly. Previously, bricks used sequential ports starting at 49152. Even though the potential port range is the same, I know that many users' firewall configurations don't allow the full range. We even say something similar in existing documentation:

https://docs.gluster.org/en/latest/Administrator%20Guide/Setting%20Up%20Clients/

> Ensure that TCP and UDP ports 24007 and 24008 are open on all Gluster servers. Apart from these ports, you need to open one port for each brick starting from port 49152 (instead of 24009 onwards as with previous releases). The brick ports assignment scheme is now compliant with IANA guidelines. For example: if you have five bricks, you need to have ports 49152 to 49156 open.

This needs to be changed and clearly said in the release notes to try to minimize connection issues after upgrade.

nik-redhat replied:
Will send out a PR for the docs too, regarding the update in the ports allocation.

nik-redhat replied:
I have created an issue for this in the docs repo https://github.com/gluster/glusterdocs/issues/609
And, once the changes are merged will send out a PR for that too.

raghavendra-talur replied:
I really like the idea of cleaning up the code around port mapper. 

However, I don't see how making the port selection random is any better than keeping it sequential but without storing stale state. As @xhernandez  has pointed out, many admins only open up as many ports in their firewall as they need. This is why max_port option was introduced recently, admin can communicate through max_port the highest port number they have open in their firewall for gluster brick. 

Comparing the situations:
1. sequential probing and no max_port set by the admin: Lesser probability of glusterd choosing a port that is blocked by the firewall.
2. random probing and no max_port set by the admin: higher probability of glusterd choosing a port that is blocked by the firewall.
3. sequential probing and max_port set by the admin: O(n) operation for glusterd to find a valid port 
4. random probing and max_port set by the admin: O(n) operation for glusterd to find a valid port

If there a use-case I have missed where random probing helps, certainly go ahead with random. Just wanted to highlight the differences. When you make the change in the doc, please explain the use of max_port as a control for firewall range.

nik-redhat replied:
@raghavendra-talur thank you for reviewing this PR, I had few things like random probing doesn't guarantee a O(n) time, but sequential probing does. 
Random probing may get into a blocked port, but that's not guaranteed. But, if we use sequential approach we will have to go through the brick ports one by one even though they are block, then isn't that costly?
Also, isn't this something similar to the magic port concept you mentioned above?

Please correct me if I am wrong somewhere.

raghavendra-talur replied:
Yes, random probing might be a little more efficient than sequential. I will leave that to you. It is not a hard requirement.


The magic port concept isn't the same but that does not block this PR either. I will create a new issue with details for that refactor. Please go ahead with this PR.

nik-redhat committed:
Added list for ports

Change-Id: I98ca1eded3dba951d9402996ad546cc1ca65d100
Signed-off-by: nik-redhat <nladha@redhat.com>

nik-redhat commented:
@xhernandez can you please run the regression tests for this PR?

xhernandez commented:
/run regression

xhernandez commented:
/run mux regression

xhernandez commented:
/run brick-mux regression

nik-redhat commented:
The test `./tests/00-geo-rep/01-georep-glusterd-tests.t` timed out and the `brick-mux-regression` failed with `No space left on device`. Can anyone please run the regression tests again for this PR?

mohit84 commented:
/run brick-mux regression

xhernandez commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run brick-mux regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run regression

nik-redhat commented:
/recheck smoke

nik-redhat commented:
/run regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run brick-mux regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run brick-mux regression

nik-redhat commented:
/recheck smoke

nik-redhat commented:
/recheck smoke

nik-redhat commented:
/run regression

nik-redhat commented:
/run brick-mux regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run brick-mux regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run brick-mux regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run brick-mux regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run brick-mux regression

nik-redhat commented:
/run brick-mux regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run brick-mux regression

nik-redhat commented:
/run brick-mux regression

nik-redhat commented:
/run brick-mux regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run brick-mux regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run regression

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-messages.h:
```diff
@@ -322,7 +322,7 @@ GLFS_MSGID(
     GD_MSG_QUORUM_CLUSTER_COUNT_GET_FAIL, GD_MSG_POST_COMMIT_OP_FAIL,
     GD_MSG_POST_COMMIT_FROM_UUID_REJCT, GD_MSG_POST_COMMIT_REQ_SEND_FAIL,
-    GD_MSG_GRACEFUL_CLEANUP_SET_FAIL);
+    GD_PMAP_PORT_BIND_FAILED, GD_MSG_GRACEFUL_CLEANUP_SET_FAIL);
```
Older messages cannot be modified. New messages must be added at the end of the list.

nik-redhat replied:
Ohh, that happened by mistake while doing a rebase....will update that.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -89,29 +88,66 @@ pmap_registry_get(xlator_t *this)
+ * the reason for the former is to ensure we don't end up with stale ports
+ */
+int
+pmap_assign_port(xlator_t *this, int old_port, char *path)
```
What prevents two concurrent calls to `pmap_assign_port()` to return the same port ?

nik-redhat replied:
I think glusterd takes care of that, the reason being we have the same mechanism being used currently and we don't face any such race condition.

xhernandez replied:
While I agree that the old implementation didn't use locks either, it had a table where assigned ports were marked so that other calls to assign a port don't return the same port. In this implementation we are not checking if a port has already been assigned.

So, if the old implementation was wrong, this new version is worse because the race window is bigger.

We need to be sure that these functions are called in a safe way or we'll need to implement some additional locking/checks here.

mykaul replied:
To me, the whole idea is to randomize. We'll then use that port. We can anyway fail using that port, because another application may have opted using it, so we'll have to go back and ask for a new, again randomized, port.

nik-redhat replied:
The old implementation wasn't wrong, rather it was using stale entries so instead of checking the ports at the startup and deciding free/in-use we are checking it only while using it.

Here, the function `pmap_port_isfree` checks for the free port and returns accordingly.....so if a brick is assigned a port number and is used by some other process before the brick process actually starts on that port then a new port is assigned to brick and it is started on the new port.

xhernandez replied:
> To me, the whole idea is to randomize. We'll then use that port. We can anyway fail using that port, because another application may have opted using it, so we'll have to go back and ask for a new, again randomized, port.

The problem is that I'm not sure if we do that. Looking at the code, a brick can be started synchronously or asynchronously. If it's started synchronously, we check for EADDRINUSE error and retry. However if it's started asynchronously, I don't see where we do that check (but maybe I've missed something).

After a quick look at the code, it seems we only start bricks synchronously after an add brick operation. Other operations seem to start them asynchronously, so probably not checking if the address already in use.

I've only done a quick look at the code, so maybe I'm wrong.

xhernandez replied:
> The old implementation wasn't wrong, rather it was using stale entries so instead of checking the ports at the startup and deciding free/in-use we are checking it only while using it.
> 
> Here, the function `pmap_port_isfree` checks for the free port and returns accordingly.....so if a brick is assigned a port number and is used by some other process before the brick process actually starts on that port then a new port is assigned to brick and it is started on the new port.

The difference is that in the previous implementation, a port just assigned to be used by a brick (besides the ports used during the initial scan) was also marked as busy, while we are not doing it now.

Anyway, as you say, as soon as a port is returned, it could become in use by another application before the brick starts, so we still need to handle this case correctly, and if we do so, it doesn't matter here.

nik-redhat replied:
>Anyway, as you say, as soon as a port is returned, it could become in use by another application before the brick starts, so we still need to handle this case correctly, and if we do so, it doesn't matter here.
 
Yeah, a check for that is present in the caller function which takes care of that part.
Like something similar happened in this bug(https://bugzilla.redhat.com/show_bug.cgi?id=1865796 ) but we have a check to prevent this scenario.

nik-redhat replied:
For starting the bricks we have the same code being used at all places which is by calling this function `glusterd_brick_start` which will handle the EADDRINUSE error and restart the bricks on new ports if we get the error.

xhernandez replied:
Yes, `glusterd_brick_start` is the common function to start the bricks, but it has an argument called `wait`. If `wait` is true, the brick is executed synchronously and the return code is checked, but if it's false, `runner_run_nowait()` is called and it doesn't seem we check if the program succeeded or not to retry with another port in case of error.

The only functions I see that pass `wait == true` are `glusterd_start_volume()` and `glusterd_op_perform_add_bricks()`, but there are several operations that use `wait == false`, so those operations can fail to start a brick if the address is in use (for example it seems that if glusterd tries to launch bricks for an already started volume, it will use `wait == false`. Snapshots also seem to have the same problem).

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -152,275 +188,185 @@ pmap_registry_search(xlator_t *this, const char *brickname,
+                   "Failed to add brick to the ports list");
+    } else {
+        tmp_brick = tmp_port->brickname;
+        asprintf(&tmp_port->brickname, "%s %s", tmp_brick, brickname);
```
It could fail.

nik-redhat replied:
Ack, I will add a check.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -152,275 +188,185 @@ pmap_registry_search(xlator_t *this, const char *brickname,
-                break;
-            }
+    pmap = pmap_registry_get(this);
+    cds_list_for_each_entry(tmp_port, &pmap->ports, port_list)
```
Instead of a list, wouldn't a small hash table be better ?

nik-redhat replied:
I thought of that, but again that would affect the working of many different parts of code along with brick-mux...so didn't continue with that approach.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -152,275 +188,185 @@ pmap_registry_search(xlator_t *this, const char *brickname,
 
-    pmap->ports[port].brickname = new_bn;
-    free(old_bn);
+    cds_list_add(&new_port->port_list, &pmap->ports);
```
Is this safe ? we don't use locks here so, unless callers ensure atomicity, we could have problems here.

nik-redhat replied:
Yeah, the locking is taken care of by the mgmt framework before reaching here.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-pmap.h:
```diff
@@ -20,38 +20,38 @@
     char *brickname;
     void *xprt;
-    gf_pmap_port_type_t type;
+    struct cds_list_head port_list;
```
Any reason for using urcu list definition instead of out own list implementation in `glusterfs/list.h` ?

nik-redhat replied:
No reason as such. As the urcu lists only were used in the glusterd component, so thought of using the same.

nik-redhat committed:
Small fixes

Change-Id: I2e9745f5f8a94d691668e5f4f59ab793327862e6
Signed-off-by: nik-redhat <nladha@redhat.com>

nik-redhat commented:
/run regression

nik-redhat committed:
Fix smoke failing for unused variables

Change-Id: I7969fc6f7031ffbbbe769e246519b7e06ba4247e
Signed-off-by: nik-redhat <nladha@redhat.com>

nik-redhat commented:
/run regression

deepshikhaaa commented:
/run regression

xhernandez reviewed (dismissed):
Based on the review comments and the state of other parts of the code, I think this patch is ok. I still think there are some race cases that are not correctly handled, but they weren't in the previous approach either, and they would probably need a bigger change outside of the scope of this one.

nik-redhat commented:
/run regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run regression

nik-redhat commented:
/run regression

xhernandez commented:
Is the documentation change also ready ? without it, users with firewalls will probably have lots of problems if not considered before upgrade.

nik-redhat commented:
> Is the documentation change also ready ? without it, users with firewalls will probably have lots of problems if not considered before upgrade.

No, it's not. I was waiting for reviews on the PR before I write the documentation.
I will write the documentation ASAP and add you to the reviewer.

nik-redhat commented:
Docs update PR: https://github.com/gluster/glusterdocs/pull/646.

Though, should we be updating the docs right away or once we have this enhancement in a new version of the gluster release?

xhernandez commented:
> Docs update PR: [gluster/glusterdocs#646](https://github.com/gluster/glusterdocs/pull/646).
> 
> Though, should we be updating the docs right away or once we have this enhancement in a new version of the gluster release?

Maybe we can say, in the documentation, that "starting from version x.y, ...". This way it can be changed now and doesn't break anything.

nik-redhat commented:
> Maybe we can say, in the documentation, that "starting from version x.y, ...". This way it can be changed now and doesn't break anything.

Okay. In that case, I will update the doc PR.

nik-redhat commented:
> Okay. In that case, I will update the doc PR.

Updated.

xhernandez commented:
@mohit84 are you also ok with this patch ?

mohit84 commented:
> @mohit84 are you also ok with this patch ?

I will check and update in next 1 or 2 days.

amarts reviewed (commented):
LGTM

+1

mohit84 commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -152,275 +188,187 @@ pmap_registry_search(xlator_t *this, const char *brickname,
+    }
+
+    if (!ret) {
+        cds_list_del_init(&tmp_port->port_list);
```
We need to cleanup brickname also.

nik-redhat replied:
If we are deleting the node from the list, won't that clear up its contents as well? Though, for the safer side, I am assigning the brickname and xprt to NULL in lines: 353, 354 before deleting the node. Shouldn't that do?

mohit84 replied:
if (!ret) {
        cds_list_del_init(&tmp_port->port_list);
        GF_FREE(tmp_port);
    }
Here you are cleanup memory for tmp_port but you are not cleanup brickname so it will become a leak.

mohit84 commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -152,275 +188,187 @@ pmap_registry_search(xlator_t *this, const char *brickname,
+                   "Failed to add brick to the ports list");
+    } else {
+        tmp_brick = tmp_port->brickname;
+        ret = gf_asprintf(&tmp_port->brickname, "%s %s", tmp_brick, brickname);
```
I think we need to delete old brickname from tmp_port before save a new brickname.

mohit84 replied:
I think new logic is not similar to previous one, we need to change this,Let's suppose we have an entry for a brickname in port list and we do call this function again it simply replace an entryname without ensuring entry is same or not.This behavior can create a big issue, we need to persist the behavior similar to older one like this
old_bn = pmap->ports[port].brickname;	
    if (old_bn) {	
        bn_len = strlen(brickname);	
        entry = strstr(old_bn, brickname);	
        while (entry) {	
            found = 1;	
            if ((entry != old_bn) && (entry[-1] != ' ')) {	
                found = 0;	
            }	
            if ((entry[bn_len] != ' ') && (entry[bn_len] != '\0')) {	
                found = 0;	
            }	
            if (found) {	
                return 0;	
            }	
            entry = strstr(entry + bn_len, brickname);	
        }	
        asprintf(&new_bn, "%s %s", old_bn, brickname);	
    } else {	
        new_bn = strdup(brickname);	
    }

mohit84 replied:
@nik-redhat We need to run regress testing to validate the patch, the portmap specific code path is sensitive and we use this to take decision about attach/detach brick so a small condition break in this code path it can create a big issue.

nik-redhat replied:
Sure.....I will update the mentioned changes, do testing of the same, and update here.

nik-redhat committed:
Added previous code checks

Change-Id: If669c9485f10a669d866da8c6766a5960e4d5754
Signed-off-by: nik-redhat <nladha@redhat.com>

nik-redhat commented:
/recheck smoke

nik-redhat commented:
/run regression

nik-redhat commented:
/recheck smoke

nik-redhat commented:
Tested the patch with create, start, stop and delete volume operations for 125 volumes with and without brick-multiplex.
And, everything was good. All the operations were successful.

nik-redhat commented:
If the patch and the test look fine, can we have the PR reviewed and merged so that it's ready for Gluster 9.1 release (doc is already updated with the same)?

mykaul commented:
> If the patch and the test look fine, can we have the PR reviewed and merged so that it's ready for Gluster 9.1 release (doc is already updated with the same)?

It should go to Gluster 10. 9.1 is a maintenance release, not a feature release.

mykaul commented on libglusterfs/src/glusterfs/mem-types.h:
```diff
@@ -120,6 +120,8 @@ enum gf_common_mem_types_ {
     gf_common_mt_server_cmdline_t,     /* used only in one location */
     gf_common_mt_latency_t,
-    gf_common_mt_end
+    gf_common_mt_pmap_reg_t,  /* used only in one location */
```
nit: I wonder why we add more memory types to the common list, while we only use it in one place. The idea is that those that are used in a single location will actually be defined there, not across all xlators.

mykaul commented on xlators/mgmt/glusterd/src/glusterd-pmap.h:
```diff
@@ -20,38 +20,39 @@
 
-struct pmap_port_status {
+struct pmap_ports {
+    int port;
```
nit: this breaks alignment and probably adds 4 bytes padding after it.

mykaul commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -152,275 +188,214 @@ pmap_registry_search(xlator_t *this, const char *brickname,
-    new_port = pmap_registry_alloc(this);
-    return new_port;
-}
+/* Allocate memory and store details about of a new port */
```
nit: rephase 'about of'

mykaul commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -467,16 +435,22 @@ gluster_pmap_portbybrick(rpcsvc_request_t *req)
 __gluster_pmap_brickbyport(rpcsvc_request_t *req)
 {
-    pmap_brick_by_port_req args = {
+    /*pmap_brick_by_port_req args = {
```
?

nik-redhat replied:
I have explained in the comment above, but instead of removing I thought of keeping it `commented` so that it can be used easily, if at all needed in the future.

nik-redhat committed:
Fix declarations and comment

Change-Id: Id630aad260a07bd165fee2c05a9d12db48233b03
Signed-off-by: nik-redhat <nladha@redhat.com>

nik-redhat commented:
/run regression

nik-redhat committed:
Replace mem type

Change-Id: If4c37707d00a1db052162a1ec176b02cdfd18940
Signed-off-by: nik-redhat <nladha@redhat.com>

nik-redhat commented:
/run regression

nik-redhat commented:
If someone can please review or merge (if it looks good), it will be helpful, it's been pending for a long time.
Thanks in advance.

Shell checks can be bypassed AFAIK.

mykaul commented on xlators/mgmt/glusterd/src/glusterd-pmap.c:
```diff
@@ -47,40 +49,37 @@ pmap_port_isfree(int port)
-    }
+    pmap->base_port = priv->base_port;
+    pmap->max_port = priv->max_port;
+    CDS_INIT_LIST_HEAD(&pmap->ports);
```
nit: init them in the order they appear in the struct.

nik-redhat committed:
Rearrange allocation of struct members

Change-Id: I8426fb7711ba099ae37717503d1e654cd3ff5c11
Signed-off-by: nik-redhat <nladha@redhat.com>

nik-redhat commented:
/run regression
